_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+Instead, we use a combination of the `commitAsync` and `commitSync` APIs to manage offset commits within the application.
+
+We use `commitAsync` for regular offset commits, allowing the consumer to continue polling for new messages without waiting for the offsets to be committed.
+
+However, if the `OffsetOutOfRangeException` or `NoOffsetForPartitionException` errors are thrown, or if we encounter an unrecoverable error, we use `commitSync` to ensure that offsets are correctly handled and committed. 
+By using the `commitSync` API, the consumer application will not poll for new messages until the last offset in the batch is committed. 
+
+**Recovering from failure**
+
+Kafka provides mechanisms to detect and recover from failures within consumer groups. Two useful properties that aid in this recovery process are `session.timeout.ms` and `heartbeat.interval.ms`.
+
+The `session.timeout.ms` property specifies the maximum amount of time a consumer in a group can go without sending a _heartbeat_ to the Kafka broker. 
+This heartbeat is an indicator that the consumer is active in the group. 
+If a consumer fails to send a heartbeat within the specified session timeout, Kafka considers it as a failure, and the consumer is marked as inactive.
+A consumer marked as inactive triggers a rebalancing of the partitions for the topic.
+Setting this value too low can result in false-positive outcomes, while setting it too high can lead to delayed recovery from failures.
+
+The `heartbeat.interval.ms` property determines how frequently a consumer sends heartbeats to the Kafka broker. 
+A shorter interval between consecutive heartbeats allows for quicker detection of consumer failures.
+The heartbeat interval must be lower, usually by a third, than the session timeout.
+Decreasing the heartbeat interval reduces the chance of accidental rebalancing, but more frequent heartbeats increases the overhead on broker resources.
+
+**Minimizing the impact of rebalances**
+
+The consumer application uses the `onPartitionsRevoked` method so that pending offsets are committed before the consumer loses ownership of the partition during a rebalance.
+
+To further reduce potential disruptions caused by rebalances, you can use the `group.instance.id` to assign a unique identifier to each consumer instance within the consumer group (as specified by `GROUP_ID` in our consumer).
+This approach minimizes unnecessary partition rebalancing when a consumer rejoins a group after a failure or restart. The consumer continues with the same instance ID and maintains its assignment of topic partitions.
+
+Additionally, adjusting the `max.poll.interval.ms` configuration can prevent rebalances caused by prolonged processing tasks, allowing you to specify the maximum interval between polls for new messages. If lengthy message processing is unavoidable due to factors such as slow external services, consider offloading this processing to a pool of worker threads. 
```

Comment:
I think this setting is more important than heartbeat.interval.ms and session.timeout.ms as it directly related to how the application is driving the consumer client.